edb/lang/edgeql/utils.py

```diff
@@ -51,21 +52,27 @@ def inline_parameters(ql_expr: qlast.Base, args: typing.Dict[str, qlast.Base]):
 
 
 def index_parameters(ql_args: typing.List[qlast.Base], *,
+                     paramnames: typing.List[str],
                      varparam: typing.Optional[int]=None):
-    result = []
-    container = result
+    result = {}
+    varargs = []
 
-    for i, e in enumerate(ql_args):
+    for (i, e), n in itertools.zip_longest(enumerate(ql_args),
+                                           paramnames,
+                                           fillvalue=None):
         if isinstance(e, qlast.SelectQuery):
             e = e.result
 
-        if i == varparam:
-            container = []
-            result.append(qlast.Array(elements=container))
+        if varparam is not None and i == varparam:
+            result[n] = qlast.Array(elements=varargs)
+            varargs.append(e)
+        else:
+            if varparam is not None and i >= varparam:
+                varargs.append(e)
+            else:
+                result[n] = e
 
```

Comment:
Maybe change `if ... else: if ... else` to `if ... elif ... else` to improve readability.